### PR TITLE
Tutorial create MicroEJ Firmware from scratch

### PR DESCRIPTION
**Pull request by gjadi-mej**

### CODE REVIEW

File: Tutorials/tutorialCreatePlatformFromScratch.rst [hunk 1]
```diff
@@ -0,0 +1,968 @@
...
+
+In WSL:
+
+#. Install qemu-arm and GNU CC toolchain for ARM: ``sudo apt-get install qemu-system-arm gcc-arm-none-eabi``
+#. The rest of this tutorial will use the folder ``src/tuto-from-scratch/`` in the Windows home folder.
+#. Create the folder: ``mkdir -p /mnt/c/Users/${USER}/src/tuto-from-scratch`` (the ``-p`` option ensures all directory are created).
+#. Go into the folder: ``cd /mnt/c/Users/${USER}/src/tuto-from-scratch/``
+#. Clone FreeRTOS and its submodules: ``git clone --recursive https://github.com/FreeRTOS/FreeRTOS.git`` (this may takes some time)
+
+Get working BSP
```
[the remark this answers is not in the document]
Comment (gjadi-mej) [reply Tutorials/tutorialCreatePlatformFromScratch.rst:74]: Ok for running.

File: Tutorials/index.rst [hunk 2]
```diff
@@ -7,6 +7,7 @@ Tutorials
    :maxdepth: 1
 
    tutorialCreateCustomPlatform
+   tutorialCreatePlatformFromScratch
```
[the remark this answers is not in the document]
Comment (gjadi-mej) [reply Tutorials/index.rst]: done

File: Tutorials/tutorialCreatePlatformFromScratch.rst [hunk 3]
```diff
@@ -0,0 +1,968 @@
+Create a MicroEJ Platform from scratch
```
[the remark this answers is not in the document]
Comment (gjadi-mej) [reply Tutorials/tutorialCreatePlatformFromScratch.rst]: done

File: Tutorials/tutorialCreatePlatformFromScratch.rst [hunk 4]
```diff
@@ -0,0 +1,968 @@
...
+======================================
+
+Introduction
+------------
+
+This tutorial explains how to create a MicroEJ Platform from scratch
```
[the remark this answers is not in the document]
Comment (gjadi-mej) [reply Tutorials/tutorialCreatePlatformFromScratch.rst]: done

File: Tutorials/tutorialCreatePlatformFromScratch.rst [hunk 5]
```diff
@@ -0,0 +1,968 @@
...
+#. Pick a target device (that meets the requirements of the project).
+#. Setup a RTOS and a toolchain that support the target device.
+#. Adapt the RTOS port if needed.
+#. Setup a MicroEJ Platform that matches the target
+   device/RTOS/toolchain.
+#. Implement Low Level API.
+#. Validate the resulting MicroEJ Platform with the `Platform
+   Qualification Tools (PQT)
+   <https://github.com/microej/PlatformQualificationTools>`_.
+#. Develop the application.
```
[the remark this answers is not in the document]
Comment (gjadi-mej) [reply Tutorials/tutorialCreatePlatformFromScratch.rst]: done

File: Tutorials/tutorialCreatePlatformFromScratch.rst [hunk 6]
```diff
@@ -0,0 +1,968 @@
...
+
+This tutorial explains how to create a MicroEJ Platform from scratch
+using QEMU, FreeRTOS and MicroEJ.
+
+The following steps are usually followed when starting a new project:
+
+#. Pick a target device (that meets the requirements of the project).
+#. Setup a RTOS and a toolchain that support the target device.
+#. Adapt the RTOS port if needed.
+#. Setup a MicroEJ Platform that matches the target
```
[the remark this answers is not in the document]
Comment (gjadi-mej) [reply Tutorials/tutorialCreatePlatformFromScratch.rst]: done

File: Tutorials/tutorialCreatePlatformFromScratch.rst [hunk 7]
```diff
@@ -0,0 +1,968 @@
...
+  <https://repository.microej.com/packages/SDK/19.05/MicroEJ-SDK-Installer-Win64-19.05.exe>`_)
+- Windows 10 with Windows Subsystem for Linux (WSL). See the
+  `installation guide
+  <https://docs.microsoft.com/en-us/windows/wsl/install-win10>`_.
+- A Linux distribution installed on WSL (Tested on Ubuntu 19.10 eoan)
+
+Overview
+--------
+
+The next steps describes step by step how to build a MicroEJ Platform
```
[the remark this answers is not in the document]
Comment (gjadi-mej) [reply Tutorials/tutorialCreatePlatformFromScratch.rst]: done

File: Tutorials/tutorialCreatePlatformFromScratch.rst [hunk 8]
```diff
@@ -0,0 +1,968 @@
...
+#. Implement the minimum Low Level API to run the application
+
+Setup the development environment
+---------------------------------
+
+This section assumes the prerequisites have been properly installed.
+
+In WSL:
+
+#. Install qemu-arm and GNU CC toolchain for ARM: ``sudo apt-get install qemu-system-arm gcc-arm-none-eabi``
```
[the remark this answers is not in the document]
Comment (gjadi-mej) [reply Tutorials/tutorialCreatePlatformFromScratch.rst]: done

File: Tutorials/tutorialCreatePlatformFromScratch.rst [hunk 9]
```diff
@@ -0,0 +1,968 @@
...
+-------------
+
+- MicroEJ SDK version 5.1.0 or higher (distribution 19.05). Can be
+  downloaded from https://developer.microej.com/ (`direct link
+  <https://repository.microej.com/packages/SDK/19.05/MicroEJ-SDK-Installer-Win64-19.05.exe>`_)
+- Windows 10 with Windows Subsystem for Linux (WSL). See the
+  `installation guide
+  <https://docs.microsoft.com/en-us/windows/wsl/install-win10>`_.
+- A Linux distribution installed on WSL (Tested on Ubuntu 19.10 eoan)
+
```
[the remark this answers is not in the document]
Comment (gjadi-mej) [reply Tutorials/tutorialCreatePlatformFromScratch.rst]: done

File: Tutorials/tutorialCreatePlatformFromScratch.rst [hunk 10]
```diff
@@ -0,0 +1,968 @@
...
+    int puts(const char *s) {
+        while (*s) {
+            putchar(*s);
+            s++;
+        }
+        return putchar('\n');
+    }
+
+And here is the patch that implements both functions and prints ``"Hello World!"``.
```
[the remark this answers is not in the document]
Comment (gjadi-mej) [reply Tutorials/tutorialCreatePlatformFromScratch.rst]: done

File: Tutorials/tutorialCreatePlatformFromScratch.rst [hunk 11]
```diff
@@ -0,0 +1,968 @@
...
+    +
+     int main( void )
+     {
+    -       puts("Hello, World!");
+    +       printf("Hello, World!\n");
+    
+            /* Configure the clocks, UART and GPIO. */
+            prvSetupHardware();
+
+At this point, the UART is implemented in the FreeRTOS BSP. The next step is to create to add the MicroEJ Platform and MicroEJ Application.
```
[the remark this answers is not in the document]
Comment (gjadi-mej) [reply Tutorials/tutorialCreatePlatformFromScratch.rst]: done

File: Tutorials/tutorialCreatePlatformFromScratch.rst [hunk 12]
```diff
@@ -0,0 +1,968 @@
...
+#. Keep the default MicroEJ Repository
+
+
+A MicroEJ Architecture is a software package that includes the MicroEJ Runtime port to a specific target Instruction Set Architecture (ISA) and C compiler. It contains a set of libraries, tools and C header files. The MicroEJ Architectures are provided by MicroEJ SDK.
+
+A MicroEJ Platform is a MicroEJ Architecture port for a custom device. It contains the MicroEJ configuration and the BSP (C source files).
+
+MicroEJ Corp. provides MicroEJ Evaluation Architectures at https://repository.microej.com/architectures/com/microej/architecture/.
+
+There is no ``CM3`` folder. This means that the MicroEJ Architectures for Cortex-M3 MCUs are no longer distributed. Download the latest MicroEJ Architecture for Cortex-M0 instead (the ARM architectures are binary upward compatible from ARMv6-M (Cortex-M0) to ARMv7-M (Cortex-M3)).
```
[the remark this answers is not in the document]
Comment (gjadi-mej) [reply Tutorials/tutorialCreatePlatformFromScratch.rst]: CM0, all screenshots have been updated.

File: Tutorials/tutorialCreatePlatformFromScratch.rst [hunk 13]
```diff
@@ -0,0 +1,968 @@
...
+    #. Accept the license and click on ``Finish``
+
+.. image:: images/tuto_microej_plat_from_scratch_import_architecture.PNG
+
+Install an Evaluation License
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+This step describes how to create and activate an Evaluation License for the MicroEJ Architecture previously imported.
+
+#. Select the ``Window > Preferences > MicroEJ > Architectures menu``.
```
[the remark this answers is not in the document]
Comment (gjadi-mej) [reply Tutorials/tutorialCreatePlatformFromScratch.rst]: what are guilabels?

File: Tutorials/tutorialCreatePlatformFromScratch.rst [hunk 14]
```diff
@@ -0,0 +1,968 @@
...
+    #. ``File > Import > MicroEJ > Architectures``
+    #. select the MicroEJ Architecture file downloaded
+    #. Accept the license and click on ``Finish``
+
+.. image:: images/tuto_microej_plat_from_scratch_import_architecture.PNG
+
+Install an Evaluation License
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+This step describes how to create and activate an Evaluation License for the MicroEJ Architecture previously imported.
```
[the remark this answers is not in the document]
Comment (gjadi-mej) [reply Tutorials/tutorialCreatePlatformFromScratch.rst]: done

File: Tutorials/tutorialCreatePlatformFromScratch.rst [hunk 15]
```diff
@@ -0,0 +1,968 @@
...
+A MicroEJ Platform is a MicroEJ Architecture port for a custom device. It contains the MicroEJ configuration and the BSP (C source files).
+
+MicroEJ Corp. provides MicroEJ Evaluation Architectures at https://repository.microej.com/architectures/com/microej/architecture/.
+
+There is no ``CM3`` folder. This means that the MicroEJ Architectures for Cortex-M3 MCUs are no longer distributed. Download the latest MicroEJ Architecture for Cortex-M0 instead (the ARM architectures are binary upward compatible from ARMv6-M (Cortex-M0) to ARMv7-M (Cortex-M3)).
+
+Import the MicroEJ Architecture
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+This step describes how to import a MicroEJ Architecture.
```
[the remark this answers is not in the document]
Comment (gjadi-mej) [reply Tutorials/tutorialCreatePlatformFromScratch.rst]: done

File: Tutorials/tutorialCreatePlatformFromScratch.rst [hunk 16]
```diff
@@ -0,0 +1,968 @@
...
+#. Go to ``Architectures`` sub-menu and check that all architectures are now activated (green check).
+#. Microej SDK is successfully activated.
+
+
+.. image:: images/tuto_microej_plat_from_scratch_activate_license.PNG
+
+Create the MicroEJ Platform
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+This step describes how to create a new MicroEJ Platform using the MicroEJ Architecture previously imported.
```
[the remark this answers is not in the document]
Comment (gjadi-mej) [reply Tutorials/tutorialCreatePlatformFromScratch.rst]: done

File: Tutorials/tutorialCreatePlatformFromScratch.rst [hunk 17]
```diff
@@ -0,0 +1,968 @@
...
+    * Set ``Name:`` to ``Tuto``
+
+.. image:: images/tuto_microej_plat_from_scratch_create_platform.PNG
+
+Setup the MicroEJ Platform
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+This step describes how to configure the MicroEJ Platform previously created.
+
+The Platform Configuration Additions provide a flexible way to configure the BSP connection between the MicroEJ Platform and MicroEJ Application to the BSP. In this tutorial, the Partial BSP connection is used. That is, the MicroEJ SDK will output all MicroEJ files (C headers, MicroEJ Application ``microejapp.o``, MicroEJ Runtime ``microejruntime.a``, ...) in a location known by the BSP. The BSP is configured to compile and link with those files.
```
[the remark this answers is not in the document]
Comment (gjadi-mej) [reply Tutorials/tutorialCreatePlatformFromScratch.rst]: done

[the remark this answers is not in the document]
Comment (gjadi-mej) [reply Tutorials/tutorialCreatePlatformFromScratch.rst]: done

File: Tutorials/tutorialCreatePlatformFromScratch.rst [hunk 18]
```diff
@@ -0,0 +1,968 @@
...
+            return 0;
+    }
+    
+    int32_t LLMJVM_IMPL_shutdown(void)
+    {
+            return LLMJVM_OK;
+    }
+
+    
+Invoke MicroEJ VM
```
[the remark this answers is not in the document]
Comment (gjadi-mej) [reply Tutorials/tutorialCreatePlatformFromScratch.rst]: done

File: Tutorials/tutorialCreatePlatformFromScratch.rst [hunk 19]
```diff
@@ -0,0 +1,968 @@
...
+~~~~~~~~~~~~~~~~~
+
+The MicroEJ VM is created and initialized with the C function ``SNI_createVM``. Then it is started and executed in the current RTOS task by calling ``SNI_startVM``. The function ``SNI_startVM`` returns when the MicroEJ Application exits. Both functions are declared in the C header ``sni.h``.
```
[the remark this answers is not in the document]
Comment (gjadi-mej) [reply Tutorials/tutorialCreatePlatformFromScratch.rst]: done

File: Tutorials/tutorialCreatePlatformFromScratch.rst [hunk 20]
```diff
@@ -0,0 +1,968 @@
...
+   int main( void )
+   {
+          printf("Hello, World!\n");
+  
+  +       SNI_startVM(SNI_createVM(), 0, NULL);
+  +
+          /* Configure the clocks, UART and GPIO. */
+          prvSetupHardware();
+
+Build and link the firmware with the MicroEJ Runtime and MicroEJ Application
```
[the remark this answers is not in the document]
Comment (gjadi-mej) [reply Tutorials/tutorialCreatePlatformFromScratch.rst]: done

File: Tutorials/tutorialCreatePlatformFromScratch.rst [hunk 21]
```diff
@@ -0,0 +1,968 @@
...
+  The MicroEJ platform library (microejruntime.a) has been deployed to: 'C:\Users\user\src\tuto-from-scratch\FreeRTOS\FreeRTOS\Demo\CORTEX_LM3S811_GCC\microej\lib'.
+  The MicroEJ platform header files (*.h) have been deployed to: 'C:\Users\user\src\tuto-from-scratch\FreeRTOS\FreeRTOS\Demo\CORTEX_LM3S811_GCC\microej\inc'.
+  =============== [ Completed Successfully ] ===============
+  
+  SUCCESS
+
+   
+At this point, the HelloWorld MicroEJ Application is built and deployed in the FreeRTOS BSP.
+
+MicroEJ and FreeRTOS integration
```
[the remark this answers is not in the document]
Comment (gjadi-mej) [reply Tutorials/tutorialCreatePlatformFromScratch.rst]: done

File: Tutorials/tutorialCreatePlatformFromScratch.rst [hunk 22]
```diff
@@ -0,0 +1,968 @@
...
+  
+  +       SNI_startVM(SNI_createVM(), 0, NULL);
+  +
+          /* Configure the clocks, UART and GPIO. */
+          prvSetupHardware();
+
+Build and link the firmware with the MicroEJ Runtime and MicroEJ Application
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+To build and link the firmware with the MicroEJ Runtime and MicroEJ Application, the FreeRTOS port must be modified to:
```
[the remark this answers is not in the document]
Comment (gjadi-mej) [reply Tutorials/tutorialCreatePlatformFromScratch.rst]: done

File: Tutorials/tutorialCreatePlatformFromScratch.rst [hunk 23]
```diff
@@ -0,0 +1,968 @@
...
+.. code-block::
+
+    CC    microej/src/LLMJVM_stub.c
+    LD    gcc/RTOSDemo.axf                                                                                                                                                                                                                     arm-none-eabi-ld: error: microej/lib/microejruntime.a(sni_vm_startup_greenthread.o) uses VFP register arguments, gcc/RTOSDemo.axf does not
+  arm-none-eabi-ld: failed to merge target specific data of file microej/lib/microejruntime.a(sni_vm_startup_greenthread.o)
+  arm-none-eabi-ld: gcc/RTOSDemo.axf section `ICETEA_HEAP' will not fit in region `SRAM'
+  arm-none-eabi-ld: region `SRAM' overflowed by 4016 bytes
+  microej/lib/microejapp.o: In function `_java_internStrings_end':   
+
+The MicroEJ Application and MicroEJ Runtime don't fit in the 8k of SRAM. It is possible to shrink the footprint and run a MicroEJ Tiny Application but this is not the purpose of this tutorial. See https://docs.microej.com/en/latest/PlatformDeveloperGuide/tiny.html for more information.
```
[the remark this answers is not in the document]
Comment (gjadi-mej) [reply Tutorials/tutorialCreatePlatformFromScratch.rst]: done

File: Tutorials/tutorialCreatePlatformFromScratch.rst [hunk 24]
```diff
@@ -0,0 +1,968 @@
...
+
+Instead, this tutorial will switch to a larger device, the Luminary Micro Stellaris LM3S6965EVB. This device has 256k of flash memory and 64k of SRAM. Update the values in the linker script ``standalone.ld``.
```
[the remark this answers is not in the document]
Comment (gjadi-mej) [reply Tutorials/tutorialCreatePlatformFromScratch.rst]: done

File: Tutorials/tutorialCreatePlatformFromScratch.rst [hunk 25]
```diff
@@ -0,0 +1,968 @@
...
+  qemu-system-arm -M lm3s6965evb -nographic -kernel gcc/RTOSDemo.bin
+
+
+.. code-block:: shell
+
+  Hello, World!
+  Hello World!
+  QEMU: Terminated // press Ctrl-a x to end the QEMU session
+
+The first ``"Hello, World!"`` is from the ``main.c`` and the second one from the MicroEJ Application.
```
[the remark this answers is not in the document]
Comment (gjadi-mej) [reply Tutorials/tutorialCreatePlatformFromScratch.rst]: done

File: Tutorials/tutorialCreatePlatformFromScratch.rst [hunk 26]
```diff
@@ -0,0 +1,968 @@
...
+
+- The MicroEJ Platform is connected to the BSP (BSP partial connection).
+- The MicroEJ Application is deployed within a known location of the BSP (in ``microej/`` folder).
+- The FreeRTOS LM3S6965 port:
+
+  * provides the minimal Low Level API to run the MicroEJ Application
+  * compiles and links FreeRTOS with the MicroEJ Application and MicroEJ Runtime
+  * runs on QEMU
+
+
```
[the remark this answers is not in the document]
Comment (gjadi-mej) [reply Tutorials/tutorialCreatePlatformFromScratch.rst]: done

File: Tutorials/tutorialCreatePlatformFromScratch.rst [hunk 27]
```diff
@@ -0,0 +1,968 @@
...
+#. Select ``File > New > MicroEJ Platform Project``.
+#. Ensure the ``Architecture`` selected is the MicroEJ Architecture previously imported.
+#. Ensure the ``Create from a platform reference implementation`` box is unchecked.
+#. Click on ``Next`` button.
+#. Fill the fields:
+
+    * Set ``Device:`` to ``lm3s811evb``
+    * Set ``Name:`` to ``Tuto``
+
+.. image:: images/tuto_microej_plat_from_scratch_create_platform.PNG
```
[the remark this answers is not in the document]
Comment (gjadi-mej) [reply Tutorials/tutorialCreatePlatformFromScratch.rst]: done

File: Tutorials/tutorialCreatePlatformFromScratch.rst [hunk 28]
```diff
@@ -0,0 +1,968 @@
...
+
+This step describes how to configure the MicroEJ Platform previously created.
+
+The Platform Configuration Additions provide a flexible way to configure the BSP connection between the MicroEJ Platform and MicroEJ Application to the BSP. In this tutorial, the Partial BSP connection is used. That is, the MicroEJ SDK will output all MicroEJ files (C headers, MicroEJ Application ``microejapp.o``, MicroEJ Runtime ``microejruntime.a``, ...) in a location known by the BSP. The BSP is configured to compile and link with those files.
+
+For this tutorial, that means that the final binary is produced by invoking ``make`` in the FreeRTOS BSP.
+
+#. Install the Platform Configuration Additions by copying all the files within the ``content`` folder in the MicroEJ Platform folder.
+
+  .. image:: images/tuto_microej_plat_from_scratch_add_platform_configuration_additions.PNG
```
[the remark this answers is not in the document]
Comment (gjadi-mej) [reply Tutorials/tutorialCreatePlatformFromScratch.rst]: done

File: Tutorials/tutorialCreatePlatformFromScratch.rst [hunk 29]
```diff
@@ -0,0 +1,968 @@
...
+In this tutorial:
+
+- The target device is a Luminary Micro Stellaris which is emulated by QEMU (LINK TO QEMU STELLARIS).
+- The RTOS and toolchain are FreeRTOS and the GNU CC toolchain for ARM.
+
+This tutorial describes step by step how to go from the FreeRTOS BSP
+to a MicroEJ Application that runs on the MicroEJ Platforms and prints
+the classic ``"Hello, World!"``.
+
+The implementation of the Low Level API and validation with the PQT
```
[the remark this answers is not in the document]
Comment (gjadi-mej) [reply Tutorials/tutorialCreatePlatformFromScratch.rst]: done

File: Tutorials/tutorialCreatePlatformFromScratch.rst [hunk 30]
```diff
@@ -0,0 +1,968 @@
+Create a MicroEJ Platform from scratch
+======================================
+
+Introduction
+------------
+
+This tutorial explains how to create a MicroEJ Platform from scratch
+using QEMU, FreeRTOS and MicroEJ.
```
[the remark this answers is not in the document]
Comment (gjadi-mej) [reply Tutorials/tutorialCreatePlatformFromScratch.rst]: @frivieremicroej Updated.

File: Tutorials/tutorialCreatePlatformFromScratch.rst [hunk 31]
```diff
@@ -0,0 +1,968 @@
...
+  microej/lib/microejapp.o: In function `__icetea__getSingleton__com_is2t_microjvm_IGreenThreadMicroJvm':
+  ... skip ...
+  microej/lib/microejapp.o: In function `TRACE_record_event_u32x3_ptr':
+  C:\Users\user\src\tuto-from-scratch\workspace\HelloWorld\com.mycompany.Main\SOAR.o:(.rodata.TRACE_record_event_u32x3_ptr+0x0): undefined reference to `TRACE_default_stub'
+  microej/lib/microejapp.o: In function `TRACE_record_event_u32x4_ptr':
+  C:\Users\user\src\tuto-from-scratch\workspace\HelloWorld\com.mycompany.Main\SOAR.o:(.rodata.TRACE_record_event_u32x4_ptr+0x0): undefined reference to `TRACE_default_stub'
+  microej/lib/microejapp.o:C:\Users\user\src\tuto-from-scratch\workspace\HelloWorld\com.mycompany.Main\SOAR.o:(.rodata.TRACE_record_event_u32x5_ptr+0x0): more undefined references to `TRACE_default_stub' follow
+  make: *** [makedefs:196: gcc/RTOSDemo.axf] Error 1
+
+This error occurs because ``microejruntime.a`` refers to symbols in ``microejapp.o``. To solve this issue, either invert the declaration of ``LIBS`` (put ``microejapp.o`` first) or guard the libraries delcaration with ``--start-group`` and ``--end-group`` in ``makedefs``. This tutorial uses the later.
```
[the remark this answers is not in the document]
Comment (gjadi-mej) [reply Tutorials/tutorialCreatePlatformFromScratch.rst]: Updated.

File: Tutorials/tutorialCreatePlatformFromScratch.rst [hunk 32]
```diff
@@ -0,0 +1,968 @@
...
+  arm-none-eabi-ld: region `SRAM' overflowed by 4016 bytes
+  microej/lib/microejapp.o: In function `_java_internStrings_end':   
+
+The MicroEJ Application and MicroEJ Runtime don't fit in the 8k of SRAM. It is possible to shrink the footprint and run a MicroEJ Tiny Application but this is not the purpose of this tutorial. See https://docs.microej.com/en/latest/PlatformDeveloperGuide/tiny.html for more information.
+
+Instead, this tutorial will switch to a larger device, the Luminary Micro Stellaris LM3S6965EVB. This device has 256k of flash memory and 64k of SRAM. Update the values in the linker script ``standalone.ld``.
+
+.. code-block:: diff
+
+  diff --git a/FreeRTOS/Demo/CORTEX_LM3S811_GCC/standalone.ld b/FreeRTOS/Demo/CORTEX_LM3S811_GCC/standalone.ld
```
[the remark this answers is not in the document]
Comment (gjadi-mej) [reply Tutorials/tutorialCreatePlatformFromScratch.rst]: Updated

File: Tutorials/tutorialCreatePlatformFromScratch.rst (hunk 13 above)
[the remark this answers is not in the document]
Comment (frivieremicroej) [reply Tutorials/tutorialCreatePlatformFromScratch.rst]: :guilabel:`File` > :guilabel:`New` > :guilabel:`Other` > :guilabel:`MicroEJ` > :guilabel:`MicroEJ Font`

[the remark this answers is not in the document]
Comment (gjadi-mej) [reply Tutorials/tutorialCreatePlatformFromScratch.rst]: done

File: Tutorials/tutorialCreateFirmwareFromScratch.rst [hunk 33]
```diff
@@ -0,0 +1,1083 @@
...
+provided by MicroEJ SDK.
+
+A MicroEJ Platform is a MicroEJ Architecture port for a custom device.
+It contains the MicroEJ configuration and the BSP (C source files).
+
+MicroEJ Corp. provides MicroEJ Evaluation Architectures at
+https://repository.microej.com/architectures/com/microej/architecture/.
+
+There is no ``CM3`` folder. This means that the MicroEJ Architectures
+for Cortex-M3 MCUs are no longer distributed. Download the latest
```
[the remark this answers is not in the document]
Comment (gjadi-mej) [reply Tutorials/tutorialCreateFirmwareFromScratch.rst]: done

File: Tutorials/tutorialCreateFirmwareFromScratch.rst [hunk 34]
```diff
@@ -0,0 +1,1083 @@
...
+<bsp_connection>` between the MicroEJ Platform and MicroEJ Application
+to the BSP. In this tutorial, the Partial BSP connection is used. That
+is, the MicroEJ SDK will output all MicroEJ files (C headers, MicroEJ
+Application ``microejapp.o``, MicroEJ Runtime ``microejruntime.a``,
+...) in a location known by the BSP. The BSP is configured to compile
+and link with those files.
+
+For this tutorial, that means that the final binary is produced by invoking ``make`` in the FreeRTOS BSP.
+
+#. Install the Platform Configuration Additions by copying all the files within the ``content`` folder in the MicroEJ Platform folder.
```
[the remark this answers is not in the document]
Comment (gjadi-mej) [reply Tutorials/tutorialCreateFirmwareFromScratch.rst]: Updated for (1) and (2).

For (3), the previous subsection links to the documentation when we talk about platform creation.

File: Tutorials/tutorialCreateFirmwareFromScratch.rst [hunk 35]
```diff
@@ -0,0 +1,1083 @@
...
+   the Board Support Package (BSP).
+#. Implement :ref:`Low Level API<LLAPI-CHAPTER>`.
+#. Validate the resulting MicroEJ Platform with the `Platform
+   Qualification Tools (PQT)
+   <https://github.com/microej/PlatformQualificationTools>`_.
+#. Develop the a :ref:`MicroEJ Application
+   <microej.sample.applications>`.
+
+This tutorial describes step by step how to go from the FreeRTOS BSP
+to a MicroEJ Application that runs on the MicroEJ Platforms and prints
```
[the remark this answers is not in the document]
Comment (gjadi-mej) [reply Tutorials/tutorialCreateFirmwareFromScratch.rst]: fixed

File: Tutorials/tutorialCreateFirmwareFromScratch.rst [hunk 36]
```diff
@@ -0,0 +1,1083 @@
...
+In this tutorial, the target device is a a Luminary Micro Stellaris.
+Though this device is no longer available on the market, it has two
+advantages:
+
+- The QEMU PC System emulator can emulate the device.
+- FreeRTOS provides an official Demo BSP.
+
+Consequently, no board is required to follow this tutorial. Everything
+is emulated on the developer's PC.
+
```
[the remark this answers is not in the document]
Comment (gjadi-mej) [reply Tutorials/tutorialCreateFirmwareFromScratch.rst:18]: Agreed. I'll ask someone to run through the tutorial and report time spent.

File: Tutorials/tutorialCreateFirmwareFromScratch.rst [hunk 37]
```diff
@@ -0,0 +1,1194 @@
...
+       linker.patch``.
+
+       It is also possible to paste the diff directly into the console:
+
+       #. In WSL, invoke ``patch -p4``. The command starts, waiting for
+          input on stdin (the standard input).
+       #. Copy the diff
+       #. Paste the diff in WSL
+       #. Press enter
+       #. Press ``Ctrl-d Ctrl-d`` (press the ``Controll`` key + the letter ``d`` twice).
```
Comment (frivieremicroej) [inline Tutorials/tutorialCreateFirmwareFromScratch.rst:179]:
```suggestion
       #. Press ``Ctrl-d Ctrl-d`` (press the ``Control`` key + the letter ``d`` twice).
```

File: Tutorials/tutorialCreateFirmwareFromScratch.rst [hunk 38]
```diff
@@ -0,0 +1,1194 @@
...
+For this tutorial, that means that the final binary is produced by
+invoking ``make`` in the FreeRTOS BSP.
+
+#. Install the Platform Configuration Additions by copying all the
+   files within the ``content``
+   `folder <https://github.com/MicroEJ/PlatformQualificationTools/tree/master/framework/platform/content>`_
+   in the MicroEJ Platform folder.
+
+   .. image:: images/tuto_microej_fw_from_scratch_add_platform_configuration_additions.PNG
+
```
Comment (frivieremicroej) [inline Tutorials/tutorialCreateFirmwareFromScratch.rst:476]: image

=> Add spacing

Comment (gjadi-mej) [reply Tutorials/tutorialCreateFirmwareFromScratch.rst:486]: This is strange. I don't have this issue on the version I built locally.

File: Tutorials/tutorialCreateFirmwareFromScratch.rst [hunk 39]
```diff
@@ -0,0 +1,1253 @@
...
+When selecting a MicroEJ Architecture, special care must be taken to
+ensure the compatibility between the toolchain used in the BSP and the
+toolchain used to build the MicroEJ Core Engine included in the
+MicroEJ Architecture.
+
+The list of MicroEJ Architectures supported is listed at
+https://docs.microej.com/en/latest/PlatformDeveloperGuide/appendix/toolchain.html.
+MicroEJ Evaluation Architectures provided by MicroEJ Corp. can be
+downloaded from `MicroEJ Artifacts Repository
```
Comment (frivieremicroej) [inline Tutorials/tutorialCreateFirmwareFromScratch.rst:377]:
```suggestion
downloaded from `MicroEJ Architectures Repository
```

File: Tutorials/tutorialCreateFirmwareFromScratch.rst [hunk 40]
```diff
@@ -0,0 +1,1253 @@
...
+<https://repository.microej.com/architectures/com/microej/architecture/>`_.
+
+There is no ``CM3`` in MicroEJ Artifacts Repository and the Arm®
```
Comment (frivieremicroej) [inline Tutorials/tutorialCreateFirmwareFromScratch.rst:380]:
```suggestion
There is no ``CM3`` in MicroEJ Architectures Repository and the Arm®
```

File: Tutorials/tutorialCreateFirmwareFromScratch.rst (hunk 36 above)
[the remark this answers is not in the document]
Comment (gjadi-mej) [reply Tutorials/tutorialCreateFirmwareFromScratch.rst:18]: Done

File: Tutorials/tutorialCreateFirmwareFromScratch.rst [hunk 41]
```diff
@@ -49,6 +49,9 @@ In this tutorial:
   <https://www.qemu.org/docs/master/system/arm/stellaris.html>`_).
 * The RTOS is FreeRTOS and the toolchain in GNU CC fo ARM.
 
+All modifications to FreeRTOS BSP made for this tutorial are available
+on https://github.com/MicroEJ/FreeRTOS/tree/tuto-microej-firmware-from-scratch.
```
Comment (frivieremicroej) [inline Tutorials/tutorialCreateFirmwareFromScratch.rst:66]:
```suggestion
at https://github.com/MicroEJ/FreeRTOS/tree/tuto-microej-firmware-from-scratch.
```

File: Tutorials/tutorialCreateFirmwareFromScratch.rst [hunk 42]
```diff
@@ -19,6 +19,19 @@ is emulated on the developer's PC.
 The tutorial should take 1 hour to complete (excluding the
 installation time of MicroEJ SDK and Windows Subsystem Linux (WSL)).
 
+Who should follow this tutorial?
```
Comment (frivieremicroej) [inline Tutorials/tutorialCreateFirmwareFromScratch.rst:22]: Intended Audience
---------------------

The audience for this document is Firmware engineers who want to understand how MicroEJ is integrated to a C Board Support Package.

**Pull request merged**